subsys/net/lib/wifi_credentials/wifi_credentials_backend_settings.c

```diff
@@ -0,0 +1,158 @@
+		return 0;
+	}
+
+	if (len < sizeof(struct wifi_credentials_header)) {
+		LOG_ERR("Settings error: invalid settings length");
+		return -EINVAL;
+	}
+
+	size_t length_read = read_cb(cb_arg, arg->buf, arg->buf_len);
+
+	/* value validation */
+	if (length_read < len) {
+		LOG_ERR("Settings error: entry incomplete");
+		return -ENODATA;
+	}
+
+	arg->found = true;
+
+	return 0;
+}
+
+/* This callback function is used to initialize the SSID cache. */
+static int zephyr_settings_backend_load_key_cb(const char *key, size_t len,
+					       settings_read_cb read_cb,
+					       void *cb_arg, void *param)
+{
+	ARG_UNUSED(param);
+
+	int ret = 0;
+
+	/* key validation */
+	if (!key) {
+		LOG_ERR("Settings error: no key");
+		return -EINVAL;
+	}
+
+	int idx = atoi(key);
+
+	if ((idx == 0 && strcmp(key, "0") != 0) || idx >= CONFIG_WIFI_CREDENTIALS_MAX_ENTRIES) {
+		LOG_ERR("Settings error: index too large");
+		return -EINVAL;
+	}
+
+	if (len < sizeof(struct wifi_credentials_header)) {
+		LOG_ERR("Settings error: invalid settings length");
+		return -EINVAL;
+	}
+
+	uint8_t buf[ENTRY_MAX_LEN];
+	size_t length_read = read_cb(cb_arg, buf, ARRAY_SIZE(buf));
+
+	/* value validation */
+	if (length_read < len) {
+		LOG_ERR("Settings error: entry incomplete");
+		return -ENODATA;
+	}
+
+	wifi_credentials_cache_ssid(idx, (struct wifi_credentials_header *)buf);
+	return 0;
+}
+
+int wifi_credentials_backend_init(void)
+{
+	int ret = settings_subsys_init();
+
+	if (ret) {
+		LOG_ERR("Initializing settings subsystem failed: %d", ret);
+		return ret;
+	}
+	ret = settings_load_subtree_direct("wifi_cred", zephyr_settings_backend_load_key_cb, NULL);
+
+	return ret;
+}
+
+int wifi_credentials_store_entry(size_t idx, const void *buf, size_t buf_len)
+{
+	char settings_name_buf[sizeof("wifi_cred/xxx")] = { 0 };
+
+	int ret = snprintk(settings_name_buf, ARRAY_SIZE(settings_name_buf), "wifi_cred/%d", idx);
+
+	if (ret < 0 || ret == ARRAY_SIZE(settings_name_buf)) {
+		LOG_ERR("WiFi credentials settings key could not be generated, idx: %d", idx);
+		return -EFAULT;
+	}
+
+	return settings_save_one(settings_name_buf, buf, buf_len);
+}
+
+int wifi_credentials_delete_entry(size_t idx)
+{
+	char settings_name_buf[sizeof("wifi_cred/xxx")] = { 0 };
+
+	int ret = snprintk(settings_name_buf, ARRAY_SIZE(settings_name_buf), "wifi_cred/%d", idx);
+
+	if (ret < 0 || ret == ARRAY_SIZE(settings_name_buf)) {
+		LOG_ERR("WiFi credentials settings key could not be generated, idx: %d", idx);
+		return -EFAULT;
+	}
+
+	return settings_delete(settings_name_buf);
+}
+
+int wifi_credentials_load_entry(size_t idx, void *buf, size_t buf_len)
+{
+	struct zephyr_settings_backend_load_cb_arg arg = {
+		.buf = buf,
+		.buf_len = buf_len,
+		.idx = idx,
+		.found = false,
+	};
+	int ret;
+
+	/* Browse through the settings entries with custom callback to load the whole entry. */
+	ret = settings_load_subtree_direct("wifi_cred", zephyr_settings_backend_load_val_cb, &arg);
```

Comment:
Yes, this iterates through the whole credentials tree. Both variants are possible, but I like the `atoi` variant more since it's robust to prefix problems. (Does the "wifi_cred/1" subtree include "wifi_cred/10"?)

Reply:
Oh, yeah, hadn't thought about that

But in that case, you could format "wifi_cred/1" as "wifi_cred/001" and this would avoid the problem you describe